File: Daily/118.pascals-triangle.js
/*
 * @lc app=leetcode id=118 lang=javascript
 *
 * [118] Pascal's Triangle
 */

// @lc code=start
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function (numRows) {
    /** @type {number[][]} */
    const triangle = [];

    for (let i = 0; i < numRows; i++) {
        const ROW_SIZE = i + 1;
        const row = new Array(ROW_SIZE);
        const prevRow = triangle[i - 1];

        row[0] = row[ROW_SIZE - 1] = 1;
        for (let j = 1; j < ROW_SIZE - 1; j++) {
            row[j] = prevRow[j - 1] + prevRow[j];
        }

        triangle[i] = row;
    }

    return triangle;
};
// @lc code=end

File: src/2025/118.pascals-triangle.cpp
/*
 * @lc app=leetcode id=118 lang=cpp
 *
 * [118] Pascal's Triangle
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<std::vector<int>> generate(const int num_rows) {
    std::vector<std::vector<int>> rows = {{1}};
    for (int i = 1; i < num_rows; i++) {
      const std::vector<int>& prev_row = rows.back();
      std::vector<int> current_row;
      current_row.reserve(prev_row.size() + 2);
      current_row.push_back(1);
      for (int j = 0; j < prev_row.size() - 1; j++) {
        current_row.push_back(prev_row[j] + prev_row[j + 1]);
      }
      current_row.push_back(1);
      rows.emplace_back(std::move(current_row));
    }

    return rows;
  }
};
// @lc code=end
